### ISSUE TITLE
[QUESTION]: Multi-task learning loss function

### ISSUE THREAD

**Issue by jamshaidsohail5** [chemprop/chemprop#398]
Hi. I hope this message finds you in the best of health. I want to know how loss function value is calculated for multiple targets while doing the training for one epoch. This below is the train function being used for one epoch and it displays a Loss  on the output for this specific epoch. If I have multiple targets, how this is being calculated ? It would be nice if you reply asap. 

```
def train(model: nn.Module,
          data: Union[MoleculeDataset, List[MoleculeDataset]],
          loss_func: Callable,
          optimizer: Optimizer,
          scheduler: _LRScheduler,
          args: Namespace,
          n_iter: int = 0,
          logger: logging.Logger = None,
          writer: SummaryWriter = None) -> int:
    """
    Trains a model for an epoch.

    :param model: Model.
    :param data: A MoleculeDataset (or a list of MoleculeDatasets if using moe).
    :param loss_func: Loss function.
    :param optimizer: An Optimizer.
    :param scheduler: A learning rate scheduler.
    :param args: Arguments.
    :param n_iter: The number of iterations (training examples) trained on so far.
    :param logger: A logger for printing intermediate results.
    :param writer: A tensorboardX SummaryWriter.
    :return: The total number of iterations (training examples) trained on so far.
    """
    debug = logger.debug if logger is not None else print
    
    model.train()
    
    data = deepcopy(data)

    data.shuffle()

    if args.uncertainty == 'bootstrap':
        data.sample(int(4 * len(data) / args.ensemble_size))

    loss_sum, iter_count = 0, 0

    num_iters = len(data) // args.batch_size * args.batch_size  # don't use the last batch if it's small, for stability

    iter_size = args.batch_size

    for i in trange(0, num_iters, iter_size):
        # Prepare batch
        if i + args.batch_size > len(data):
            break
        mol_batch = MoleculeDataset(data[i:i + args.batch_size])
        smiles_batch, features_batch, target_batch = mol_batch.smiles(), mol_batch.features(), mol_batch.targets()
        batch = smiles_batch
        mask = torch.Tensor([[x is not None for x in tb] for tb in target_batch])
        targets = torch.Tensor([[0 if x is None else x for x in tb] for tb in target_batch])

        if next(model.parameters()).is_cuda:
            mask, targets = mask.cuda(), targets.cuda()

        class_weights = torch.ones(targets.shape)

        if args.cuda:
            class_weights = class_weights.cuda()

        # Run model
        model.zero_grad()
        preds = model(batch, features_batch)

        if model.uncertainty:
            pred_targets = preds[:, [j for j in range(len(preds[0])) if j % 2 == 0]]
            pred_var = preds[:, [j for j in range(len(preds[0])) if j % 2 == 1]]
            loss = loss_func(pred_targets, pred_var, targets)
            # sigma = ((pred_targets - targets) ** 2).detach()
            # loss = loss_func(pred_targets, targets) * class_weights * mask
            # loss += nn.MSELoss(reduction='none')(pred_sigma, sigma) * class_weights * mask
        else:
            loss = loss_func(preds, targets) * class_weights * mask
        loss = loss.sum() / mask.sum()

        loss_sum += loss.item()
        iter_count += len(mol_batch)

        loss.backward()
        optimizer.step()

        if isinstance(scheduler, NoamLR):
            scheduler.step()

        n_iter += len(mol_batch)

        # Log and/or add to tensorboard
        if (n_iter // args.batch_size) % args.log_frequency == 0:
            lrs = scheduler.get_lr()
            pnorm = compute_pnorm(model)
            gnorm = compute_gnorm(model)
            loss_avg = loss_sum / iter_count
            loss_sum, iter_count = 0, 0

            lrs_str = ', '.join(f'lr_{i} = {lr:.4e}' for i, lr in enumerate(lrs))
            debug(f'Loss = {loss_avg:.4e}, PNorm = {pnorm:.4f}, GNorm = {gnorm:.4f}, {lrs_str}')

            if writer is not None:
                writer.add_scalar('train_loss', loss_avg, n_iter)
                writer.add_scalar('param_norm', pnorm, n_iter)
                writer.add_scalar('gradient_norm', gnorm, n_iter)
                for i, lr in enumerate(lrs):
                    writer.add_scalar(f'learning_rate_{i}', lr, n_iter)

    return n_iter

```

Best Wishes
Jamshaid Sohail

**Comment by cjmcgill** [comment]
A minor correction @kspieks , it was only the metrics calculated on the test set and validation set that we changed to being the geometric mean. It mattered because those were calculated using unscaled targets and so were magnitude-dependent. Inside the model during training, the targets are scaled to a zscore, so different orders of magnitude no longer matter, so arithmetic mean is fine inside training.

The specific line in this code where the averaging happens is `loss = loss.sum() / mask.sum()`. This takes in the loss for all the data in a minibatch for all the targets, a tensor of shape (batch_size x number_of_targets). It takes the sum of that loss and divides it by the total number of target entries (`mask.sum()`), excluding skipped targets. So it's an arithmetic mean on the basis of individual targets in a batch. It skips over blank entries though where you don't provide a target, so if you have blanks disproportionately in one target then the loss will be weighted toward the targets without blanks. This is all done relative to scaled targets, so your units or order of magnitude in target values shouldn't matter.

**Comment by kspieks** [comment]
Thanks for clarifying with this additional detail! Yes training targets are z-scored so the type of mean is important for val and test
